lib/fencing/st_client.c

```diff
@@ -1730,41 +1730,22 @@ stonith_api_delete(stonith_t * stonith)
     }
 }
 
-static int
-stonith_api_validate(stonith_t *st, int call_options, const char *rsc_id,
-                     const char *namespace_s, const char *agent,
-                     const stonith_key_value_t *params, int timeout_sec,
-                     char **output, char **error_output)
+int
+stonith__validate(stonith_t *st, int call_options, const char *rsc_id,
+                  const char *namespace_s, const char *agent,
+                  GHashTable *params, const char *host_arg, int timeout_sec,
+                  char **output, char **error_output)
 {
-    /* Validation should be done directly via the agent, so we can get it from
-     * stonith_admin when the cluster is not running, which is important for
-     * higher-level tools.
-     */
-
-    int rc = pcmk_ok;
+    int rc = pcmk_rc_ok;
 
     /* Use a dummy node name in case the agent requires a target. We assume the
      * actual target doesn't matter for validation purposes (if in practice,
      * that is incorrect, we will need to allow the caller to pass the target).
      */
     const char *target = "node1";
-    const char *host_arg = NULL;
-
-    GHashTable *params_table = pcmk__strkey_table(free, free);
-
-    // Convert parameter list to a hash table
-    for (; params; params = params->next) {
-        if (pcmk__str_eq(params->key, PCMK_STONITH_HOST_ARGUMENT,
-                         pcmk__str_none)) {
-            host_arg = params->value;
-        }
-        if (!pcmk_stonith_param(params->key)) {
```

Comment:
The previous code would add only if `!pcmk_stonith_param()`. The new code will have to remove the stonith params. (Add a doxygen block noting params as in,out.)

Reply:
I'm not quite following what you mean here by needing to remove the stonith params.

Reply:
The previous code didn't copy all input parameters to the list, only agent parameters (`pcmk_stonith_param()` detects special parameters interpreted by Pacemaker instead of the agent, so those must not be passed to the agent.)